lib/punchblock/connection/asterisk.rb

```diff
@@ -38,34 +39,69 @@ def handle_event(event)
       end
 
       def new_ami_stream
-        stream = RubyAMI::Stream.new(*@stream_options, ->(event) { translator.async.handle_ami_event event }, pb_logger)
-        client = (ami_client || RubyAMIStreamProxy.new(stream))
-        client.stream = stream
-        client
+        Celluloid::Actor[:ami_stream] = RubyAMI::Stream.new(*@stream_options, ->(event) { translator.async.handle_ami_event event }, pb_logger)
+        ActorHandle.new :ami_stream
       end
 
       def start_ami_client
         @ami_client = new_ami_stream unless ami_client.alive?
         ami_client.async.run
-        Celluloid::Actor.join(ami_client)
+        Celluloid::Actor.join Celluloid::Actor[:ami_stream]
       end
 
       def new_call_uri
         Punchblock.new_uuid
       end
     end
 
-    class RubyAMIStreamProxy
-      attr_accessor :stream
-
-      def initialize(ami)
-        @stream = ami
+    # Provides a handle to an Actor that can safely be cached/memoized without
+    # risk of that handle becoming stale from a DeadActorError.
+    #
+    # * Requirement *:
+    # The Actor must be put into the Celluloid::Actor Registry, which you are
+    # responsible for setting up yourself.
+    # @see https://github.com/celluloid/celluloid/wiki/Registry
+    # @see https://github.com/celluloid/celluloid/wiki/Supervisors
+    #
+    # Example:
+    #   require 'celluloid'
+    #   class Clumsy
+    #     include Celluloid
+    #     def initialize
+    #       puts "Clumsy ID badge # #{object_id} reporting for duty, sir!"
+    #     end
+    #   end
+    #   supervisor = Clumsy.supervise_as :clumsy
+    #   # Output: Clumsy ID badge # 2352 reporting for duty, sir!
+    #
+    #   clumsy_stale = Celluloid::Actor[:clumsy] # Don't cache clumsy_stale, he will eventually break!
+    #   clumsy_fresh = ActorHandle.new :clumsy   # Use this instead!
+    #
+    #   clumsy_fresh.async.send :no_method_kaboom # Oh no! A DeadActor!
+    #   # Output: NoMethodError: undefined method `no_method_kaboom' for #<Celluloid::ActorProxy(Clumsy:0x930)
+    #   # Output: Clumsy ID badge # 2356 reporting for duty, sir!
+    #   p clumsy_stale.alive? # Dead :(
+    #   p clumsy_fresh.alive? # Not Dead! =)
+    class ActorHandle
+      def initialize(registered_name)
+        @_registered_name = registered_name
+        fallback_handle
       end
 
       def method_missing(method, *args, &block)
-        stream.__send__(method, *args, &block)
+        fallback_handle.__send__(method, *args, &block)
       end
 
+      # Maintain a fallback handle to a Dead Actor.  Why?
```

Comment:
love the explanation - let's ditch one of the `Why?`s or change it to `Why, oh why?` 😉 

otherwise, 👍